src/inmanta/env.py

```diff
@@ -69,17 +124,106 @@ def install_from_source(cls, paths: List[LocalPackagePath]) -> None:
+        :param constraints: In addition to checking for compatibility within the environment, also verify that the environment's
+            packages meet the given constraints. This is insterpreted as a list of constraints, not a list of requirements: if
+            one of the constraints' package is not installed the check succeeds.
+        """
+
+        def check_installed() -> List[str]:
+            process: subprocess.CompletedProcess = subprocess.run(
+                [
+                    cls.python_path,
+                    "-m",
+                    "pip",
+                    "check",
+                ],
+                stdout=subprocess.PIPE,
+            )
+            # returncodes 0 and 1 are expected
+            if process.returncode == 0:
+                return []
+            elif process.returncode == 1:
+                stdout: str = process.stdout.decode()
+                return [line for line in stdout.splitlines() if in_scope.fullmatch(line.split()[0])]
+            else:
+                process.check_returncode()
+            return []
+
+        def check_constraints() -> Iterator[Tuple[Requirement, version.Version]]:
+            if constraints is None:
+                return iter(())
+
+            installed_packages: Dict[str, version.Version] = {
+                name.lower(): version for name, version in get_installed_packages(cls.python_path).items()
+            }
+            return (
+                (constraint, installed_packages[constraint.key])
+                for constraint in constraints
+                if constraint.key in installed_packages and str(installed_packages[constraint.key]) not in constraint
+            )
+
+        incompatibilities: List[str] = check_installed()
+        for incompatibility in incompatibilities:
+            LOGGER.warning("Incompatibility in Python environment: {incompatibility}")
+        constraint_violations: List[Tuple[Requirement, version.Version]] = list(check_constraints())
+        for constraint, v in constraint_violations:
+            LOGGER.warning(f"Incompatibility between constraint {constraint} and installed version {v}")
+
+        return len(incompatibilities) == 0 and len(constraint_violations) == 0
+
+    @classmethod
+    def get_module_file(cls, module: str) -> Optional[Tuple[str, Loader]]:
+        """
+        Get the location of the init file for a Python module within the active environment.
+
+        :return: A tuple of the path and the associated loader, if the module is found.
+        """
+        spec: Optional[ModuleSpec]
+        try:
+            spec = importlib.util.find_spec(module)
+        # inmanta.loader.PluginModuleLoader raises ImportError if module is not found
+        except (ImportError, ModuleNotFoundError):
+            spec = None
+        # TODO: test with v1 loader
+        return (spec.origin, spec.loader) if spec is not None else None
+
+    @classmethod
+    def init_namespace(cls, namespace: str) -> None:
```

Comment:
I think @arnaudsjs solved the problem of the namespace by adding it to the core package

Reply:
This was definitely causing me issues (actually because it was included in the core package, if I removed it the tests succeeded). The issue is that one `importlib` loads from the namespace once, it caches the search locations (`importlib.invalidate_caches()` doesn't seem to affect this) so on a subsequent call, if a new path exists, it doesn't get detected. Only the old path is then searched for namespace packages.

Reply:
also with the additional reset calls that @arnaudsjs added? He did figure out how to reload the namespaces iirc?

Reply:
Yes, I merged in master and some tests started to fail. When I deleted `src/inmanta_plugins/` they succeeded again. Creating the other directory before loading the package the first time (what this method does) solved it.